src/main/java/io/numaproj/numaflow/function/FunctionService.java

```diff
@@ -3,64 +3,165 @@
+
+    private static final Logger logger = Logger.getLogger(FunctionService.class.getName());
+    private MapHandler mapHandler;
+    private ReduceHandler reduceHandler;
+
+    private StreamObserver<Udfunction.Datum> streamObserver;
+
+    public FunctionService() {
+    }
+
+    public void setMapHandler(MapHandler mapHandler) {
+        this.mapHandler = mapHandler;
+    }
+
+    public void setReduceHandler(ReduceHandler reduceHandler) {
+        this.reduceHandler = reduceHandler;
+    }
+
+    /**
+     * Applies a function to each datum element.
+     */
+    @Override
+    public void mapFn(Udfunction.Datum request, StreamObserver<Udfunction.DatumList> responseObserver) {
+        if (this.mapHandler == null) {
+            io.grpc.stub.ServerCalls.asyncUnimplementedUnaryCall(getMapFnMethod(), responseObserver);
+            return;
+        }
+
+        // get key from gPRC metadata
+        String key = Function.DATUM_CONTEXT_KEY.get();
+
+        // get Datum from request
+        Udfunction.Datum handlerDatum = Udfunction.Datum.newBuilder()
+                .setValue(request.getValue())
+                .setEventTime(request.getEventTime())
+                .setWatermark(request.getWatermark())
+                .build();
+
+        // process Datum
+        Message[] messages = mapHandler.HandleDo(key, handlerDatum);
+
+        // set response
+        responseObserver.onNext(buildDatumList(messages));
+        responseObserver.onCompleted();
     }
 
-    // get key from gPRC metadata
-    String key = Function.DATUM_CONTEXT_KEY.get();
+    /**
+     * Streams input data to reduceFn and returns the result.
+     */
+    @Override
+    public StreamObserver<Udfunction.Datum> reduceFn(StreamObserver<Udfunction.DatumList> responseObserver) {
+        if (this.reduceHandler == null) {
+            return io.grpc.stub.ServerCalls.asyncUnimplementedStreamingCall(getReduceFnMethod(), responseObserver);
+        }
+
+        ReduceDatumStreamImpl reduceDatumStreamImpl = new ReduceDatumStreamImpl();
+
+        // get key from gPRC metadata
+        String key = Function.DATUM_CONTEXT_KEY.get();
+
+        // get window start and end time from gPRC metadata
+        String winSt = Function.WINDOW_START_TIME.get();
+        String winEt = Function.WINDOW_END_TIME.get();
+
+        // convert the start and end time to Instant
+        Instant startTime = Instant.ofEpochMilli(Long.parseLong(winSt));
+        Instant endTime = Instant.ofEpochMilli(Long.parseLong(winEt));
+
+        // create metadata
+        IntervalWindow iw = new IntervalWindowImpl(startTime, endTime);
+        Metadata md = new MetadataImpl(iw);
+
 
-    // get Datum from request
-    Udfunction.Datum handlerDatum = Udfunction.Datum.newBuilder()
-        .setValue(request.getValue())
-        .setEventTime(request.getEventTime())
-        .setWatermark(request.getWatermark())
-        .build();
+        ExecutorService executor = Executors.newSingleThreadExecutor();
```

Comment:
consider adding a shutdown hook that can cancel the tasks. 

Also consider using simple Callable with FutureTask. although the incidence of too many executors is rare, we have seen deadlocks with too many executors.

Reply:
added fixed thread pool